packages/zk-eddsa-event-ticket-pcd/src/ZKEdDSAEventTicketPCD.ts

```diff
@@ -0,0 +1,538 @@
+
+  // `fieldsToReveal`, `externalNullifier`, `watermark` are usually app-specified
+  fieldsToReveal: ObjectArgument<EdDSATicketFieldsToReveal>;
+  watermark: BigIntArgument;
+
+  // provide externalNullifier field to request a nullifierHash
+  // if you don't provide this field, no nullifierHash will be outputted
+  externalNullifier?: BigIntArgument;
+}
+
+/**
+ * Claim part of a ZKEdDSAEventTicketPCD contains all public/revealed fields.
+ */
+export interface ZKEdDSAEventTicketPCDClaim {
+  partialTicket: Partial<ITicketData>;
+  watermark: string;
+  signer: EDdSAPublicKey; // in montgomery form. must use F.toObject() from ffjavascript to convert to raw coords
+
+  // only if requested in PCDArgs
+  validEventIds?: string[];
+  externalNullifier?: string;
+  nullifierHash?: string;
+}
+
+/**
+ * Proof part of a ZKEdDSAEventTicketPCD is a snarkjs proof.
+ */
+export interface ZKEdDSAEventTicketPCDProof {
+  // TODO: Can this be imported from somewhere?
+  pi_a: [string, string, string];
+  pi_b: [[string, string], [string, string], [string, string]];
+  pi_c: [string, string, string];
+  protocol: string;
+  curve: string;
+}
+
+/**
+ * ZKEdDSAEventTicketPCD PCD type representation.
+ */
+export class ZKEdDSAEventTicketPCD
+  implements PCD<ZKEdDSAEventTicketPCDClaim, ZKEdDSAEventTicketPCDProof>
+{
+  type = ZKEdDSAEventTicketPCDTypeName;
+
+  public constructor(
+    readonly id: string,
+    readonly claim: ZKEdDSAEventTicketPCDClaim,
+    readonly proof: ZKEdDSAEventTicketPCDProof
+  ) {
+    this.id = id;
+    this.claim = claim;
+    this.proof = proof;
+  }
+}
+
+/**
+ * Initialize ZKEdDSAEventTicketPCDPackage.
+ */
+export async function init(args: ZKEdDSAEventTicketPCDInitArgs) {
+  initArgs = args;
+}
+
+async function ensureInitialized() {
+  if (!initArgs) {
+    throw new Error(
+      "Cannot initialize ZKEdDSAEventTicketPCDPackage: init has not been called yet"
+    );
+  }
+
+  if (!initializedPromise) {
+    initializedPromise = (async () => {
+      babyJub = await buildBabyjub();
+      eddsa = await buildEddsa();
+    })();
+  }
+
+  await initializedPromise;
+}
+
+async function checkProveInputs(args: ZKEdDSAEventTicketPCDArgs): Promise<{
+  ticketPCD: EdDSATicketPCD;
+  identityPCD: SemaphoreIdentityPCD;
+  fieldsToReveal: EdDSATicketFieldsToReveal;
+  watermark: bigint;
+}> {
+  const serializedTicketPCD = args.ticket.value?.pcd;
+  if (!serializedTicketPCD) {
+    throw new Error("Cannot make proof: missing ticket PCD");
+  }
+
+  const serializedIdentityPCD = args.identity.value?.pcd;
+  if (!serializedIdentityPCD) {
+    throw new Error("Cannot make proof: missing identity PCD");
+  }
+
+  const fieldsToReveal = args.fieldsToReveal.value;
+  if (!fieldsToReveal) {
+    throw new Error("Cannot make proof: missing fields request object");
+  }
+
+  if (!args.watermark.value) {
+    throw new Error("Cannot make proof: missing watermark");
+  }
+
+  if (
+    args.externalNullifier !== undefined &&
```

Comment:
once you make `externalNullifier` not be possibly undefined, this condition will be superfluous